### PR TITLE
IOS - add initial set of contracts and validations

### PR DESCRIPTION
**Pull request by mdkaraman**
ME-182215348

**Note**
This is the first PR for work related to MaineCare2.0, which involves implementing the new IOS (Integrated Online Services) payload schema for outbound account transfers.  This will require a brand new set of contracts, validations and operations which are presently not called from any existing service in the IC ecosystem.  As such, there is no risk of breaking changes or other conflicts that can occur when merging new code into a shared component like aca_entities.  All PR's related to this work will be prefixed with "IOS" for easy identification.

### CODE REVIEW

**Comment by kristinmerbach** [comment]
All initial contracts, specs and validations are added, should be good to start the review

File: lib/aca_entities/medicaid/ios/contracts/contact_contract.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        # Contract for Contact
+        class ContactContract < Dry::Validation::Contract
+          params do
+            optional(:AddressTypeCode__c).maybe(:string)
+            optional(:AgencyEmployeeId__c).maybe(:integer)
+            optional(:AssociationStartDate__c).maybe(:date)
+            optional(:Birthdate).maybe(:date)
+            optional(:Email).maybe(:string)
+            required(:FirstName).filled(:string)
+            required(:LastName).filled(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/contact_contract.rb:16]: "last name" doesn't appear on the "requeried fields" tab

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            optional(:DependentCareProvider__c).maybe(:string)
+            optional(:DependentCareProviderId).maybe(:integer)
+            optional(:DependentIndividual__c).maybe(:string)
+            optional(:EIN__c).maybe(:integer)
+            optional(:EmployerName__c).maybe(:string)
+            optional(:EmploymentEndReason__c).maybe(:string)
+            optional(:EndDate__c).maybe(:date)
+            optional(:ExpenseAmount__c).maybe(:float)
+            optional(:ExpenseFrequencyCode__c).maybe(:string)
+            optional(:ExpenseSubType__c).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:50]: this is marked as required on the database, shall we set as required here ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            optional(:DCSelfEmploymentIncomeDetailId__c).maybe(:integer)
+            optional(:DCUnearnedIncomeDetailId__c).maybe(:integer)
+            optional(:DependentCareProvider__c).maybe(:string)
+            optional(:DependentCareProviderId).maybe(:integer)
+            optional(:DependentIndividual__c).maybe(:string)
+            optional(:EIN__c).maybe(:integer)
+            optional(:EmployerName__c).maybe(:string)
+            optional(:EmploymentEndReason__c).maybe(:string)
+            optional(:EndDate__c).maybe(:date)
+            optional(:ExpenseAmount__c).maybe(:float)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:48]: this is marked as required on the database, shall we set as required here ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            optional(:ExpenseFrequencyCode__c).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:49]: this is marked as required on the database, shall we set as required here ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+            optional(:ExpenseSubType__c).maybe(:string)
+            optional(:ExpenseTypeCode__c).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:51]: this is marked as required on the database, shall we set as required here ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            optional(:FaceValueAmount__c).maybe(:integer)
+            optional(:FeesAmount__c).maybe(:integer)
+            optional(:FuneralFundCode__c).maybe(:string)
+            optional(:FuneralLocation__c).maybe(:string)
+            optional(:HasItemizedStmtOfGoodsSignByClientToggle__c).maybe(:string)
+            optional(:HasPolicyAssignedToFuneralHomeToggle__c).maybe(:string)
+            optional(:IncomePayDetailHoursPerWeek__c).maybe(:integer)
+            optional(:IncomePayFrequency__c).maybe(:string)
+            optional(:IncomeTypeCode__c).maybe(:string)
+            optional(:IncomeSubtypeCode__c).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:61]: this is marked as required on the database, shall we set as required here ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+            optional(:PhysicalCity__c).maybe(:string)
+            optional(:PhysicalStateCode__c).maybe(:string)
+            optional(:PhysicalZipCode5__c).maybe(:string)
+            required(:PreferredNotificationMethodCode__c).filled(:string)
+            required(:TaxFilerMemberCurrent__r).filled(:hash)
+            required(:LastName__c).filled(:string)
+            required(:MemberProgramsApplied__c).filled(:string)
+            required(:IsFederalRecognizedIndianTribeToggle__c).filled(:string)
+            required(:IsPregnant__c).filled(:string)
+            optional(:Name).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb:23]: we have first name and last name, is this necessary ?

File: lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+            optional(:MailingAddrValidated__c).maybe(:string)
+            optional(:PhysicalCountyCode__c).maybe(:string)
+            optional(:PhysicalZipCode4__c).maybe(:string)
+            optional(:MailingAddressLine1__c).maybe(:string)
+            optional(:MailingAddressLine2__c).maybe(:string)
+            optional(:MailingCity__c).maybe(:string)
+            optional(:MailingCountyCode__c).maybe(:string)
+            optional(:MailingStateCode__c).maybe(:string)
+            optional(:MailingZipCode4__c).maybe(:string)
+            optional(:MailingZipCode5__c).maybe(:string)
```
Comment (polographer) [inline lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb:166]: this is marked as required on the database, shall we set as required here ?

**Review by scaustin34 — CHANGES_REQUESTED** [review]

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb (hunk 4 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:49]: It is required in the database if the table being built is an expense. the asset model also includes incomes. Propose instead adding a validation that it is required if it has an expense type.

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb (hunk 5 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:51]: t is required in the database if the table being built is an expense. the asset model also includes incomes. Propose instead adding a validation that it is required conditionally, though still need to work on the logic for this one.

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb (hunk 6 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:61]: Similar to the expenses, but going the other way. I'll add in some conditional validation

File: lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb (hunk 7 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb:23]: Will throw this on the questions for the client list. assuming it is similar to enroll having a full name field for ease of reference

File: lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb (hunk 8 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb:166]: These are required if you're building a full address, we wouldn't be doing so if they had the same mailing and physical address. I'll add conditional validations to that effect

File: lib/aca_entities/medicaid/ios/contracts/ssp_application__c_contract.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module AcaEntities
+  module Medicaid
+    module Ios
+      module Contracts
+        # Contract for SSP_Application__c
+        class SspApplicationCContract < Dry::Validation::Contract
+          params do
```
[the remark this answers is not in the document]
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_application__c_contract.rb:9]: adding the fields, the validation requires info from fields not in this contract, so cannot be added here

File: lib/aca_entities/medicaid/ios/contracts/ssp_insurance_policy__c_contract.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    module Ios
+      module Contracts
+        # Contract for SSP_InsurancePolicy__c
+        class SspInsurancePolicyCContract < Dry::Validation::Contract
+          params do
+            optional(:DCId__c).maybe(:integer)
+            required(:Is_this_an_empl__c).filled(:string)
+            optional(:InsuranceCompanyName__c).maybe(:string)
+            optional(:InsuranceGroupNumber__c).maybe(:string)
+            optional(:InsurancePolicyNumber__c).maybe(:string)
```
[the remark this answers is not in the document]
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_insurance_policy__c_contract.rb:14]: added!

File: lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb (hunk 8 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb:166]: Actually, I already wrote those validations, they're [already in the code base](lib/aca_entities/medicaid/ios/contracts/ssp_member__c_contract.rb) lines 660-693.

File: lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb (hunk 5 above)
Comment (kristinmerbach) [reply lib/aca_entities/medicaid/ios/contracts/ssp_asset__c_contract.rb:52]: Added validation is there is an expense amount, the type cannot be blank

**Review by polographer — APPROVED** [review]
looks good !

**Review by scaustin34 — APPROVED** [review]
Looks good!

**Pull request merged**